lightning/src/util/persist.rs

```diff
@@ -243,10 +250,889 @@ where
 			Err(_) => {
 				return Err(io::Error::new(
 					io::ErrorKind::InvalidData,
-					"Failed to deserialize ChannelMonitor"
+					"Failed to read ChannelMonitor"
 				))
 			}
 		}
 	}
 	Ok(res)
 }
+
+/// Implements [`Persist`] in a way that writes and reads both [`ChannelMonitor`]s and
+/// [`ChannelMonitorUpdate`]s.
+///
+/// # Overview
+///
+/// The main benefit this provides over the [`KVStore`]'s [`Persist`] implementation is decreased
+/// I/O bandwidth and storage churn, at the expense of more IOPS (including listing, reading, and
+/// deleting) and complexity. This is because it writes channel monitor differential updates,
+/// whereas the other (default) implementation rewrites the entire monitor on each update. For
+/// routing nodes, updates can happen many times per second to a channel, and monitors can be tens
+/// of MB (or more). Updates can be as small as a few hundred bytes.
+///
+/// Note that monitors written with this struct are _not_ backward-compatible with the default
+/// [`KVStore`]'s [`Persist`] implementation. They have a prepended byte sequence,
+/// [`MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL`], applied to prevent deserialization with other
+/// persisters. This is because monitors written by this struct _may_ have unapplied updates,
+/// particularly if the node panics and aborts. In order to downgrade, you must ensure that all
+/// updates are applied to the monitor, and remove the sentinel bytes.
+///
+/// # Storing monitors
+///
+/// Monitors are stored by implementing the [`Persist`] trait, which has two functions:
+///
+///   - [`Persist::persist_new_channel`], which persists whole [`ChannelMonitor`]s.
+///   - [`Persist::update_persisted_channel`], which persists only a [`ChannelMonitorUpdate`]
+///
+/// Whole [`ChannelMonitor`]s are stored in the [`CHANNEL_MONITOR_PERSISTENCE_NAMESPACE`], using the
+/// familiar encoding of an [`OutPoint`]'s transaction ID and index (for example,
+/// `[SOME-64-CHAR-HEX-STRING]_1`).
+///
+/// Each [`ChannelMonitorUpdate`] is stored in a dynamic sub-namespace, as follows:
+///
+///   - namespace: [`CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE`]
+///   - sub-namespace: [the monitor's encoded outpoint name]
+///
+/// Under that namespace, each update is stored with a left-padded `u64`, like
```

Comment:
This has all changed so much that it may not be helpful anymore.

When we used to list updates, we also had to make sure that we processed the items in the list in numerical (_not_ lexicographic) order. So initially I did a resource-intensive thing by just deserializing all the updates and then sorting them in-memory using their update_ids, essentially not trusting the name at all. Then we moved to not doing that and just sorting the listing (you may recall, for a minute `KVStore::list` had a lex sort requirement, too). So this was to ensure that listings in lex sort order would also be in numerical order, and we wouldn't need to sort. I think once we removed that requirement from `KVStore::list` I was still sorting, but most storage would return the list lexicographically by default, and I think sorted list is a best case for timsort, so it was all good.

_<takes breath>_

But! Now, this may actually be obviated. We don't list anymore. We are going to add listing back, but only for the "deep clean" which doesn't really _need_ the updates to be listed in numerical order (maybe? I'm not sure, it might be more efficient that way). Beyond that, I think it only comes up when the customer lists updates in some kind of troubleshooting scenario. Would it be better to have the updates just be the integer, or to have them always list in numerical order?

I suppose if we're neutral about all of the above, less code is better and I can get rid of this now.

Reply:
Thinking on it more -- let's revisit after we add listing back for deep cleans.

Reply:
Okay, left padding is removed. There's an optimization we could do in `clean_stale_updates` by sorting, but I don't think it's terribly important to worry about right now so I will save it for another PR.